opencog/guile/SchemePrimitive.h

```diff
@@ -30,640 +78,231 @@ class PrimitiveEnviron
-	SchemePrimitive(const char* module, const char* name, \
-		 RET_TYPE (T::*cb)(void), T *data) \
-	{ \
-		that = data; \
-		method.LSIG = cb; \
-		scheme_module = module; \
-		scheme_name = name; \
-		signature = SIG; \
-		do_register(module, name, 0); /* cb has 0 args */ \
+	typedef R (C::*Method)(Args...);
+	using PlainTuple = std::tuple<typename std::remove_reference<Args>::type...>;
+
+	Method method;
+	C* that;
+	const char *scheme_module;
+	const char *scheme_name;
+
+public:
+	SchemePrimitiveBase(const char* module, const char* name,
+	                    R (C::*cb)(Args...), C *data)
+	{
+		that = data;
+		method = cb;
+		scheme_module = module;
+		scheme_name = name;
+		do_register(module, name, sizeof...(Args));
 	}
 
-#define DECLARE_CONSTR_1(SIG, LSIG, RET_TYPE, ARG_TYPE) \
-	SchemePrimitive(const char* module, const char* name, \
-		 RET_TYPE (T::*cb)(ARG_TYPE), T *data) \
-	{ \
-		that = data; \
-		method.LSIG = cb; \
-		scheme_module = module; \
-		scheme_name = name; \
-		signature = SIG; \
-		do_register(module, name, 1); /* cb has 1 arg */ \
+protected:
+	virtual const char *get_name(void) { return scheme_name; }
+	virtual const char *get_module(void) { return scheme_module; }
+	virtual size_t get_size(void) { return sizeof (*this); }
+
+	// Convert the ith scm argument to its C++ object. The reason the
+	// return type is also present as third argument is to tell gcc
+	// what overloaded version of scm_to to use. It is a bit of a hack
+	// but not that ugly given that the alternative is to use
+	// specialized templates.
+	bool scm_to(SCM args, size_t idx, bool) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return scm_to_bool(arg);
+	}
+	size_t scm_to(SCM args, size_t idx, size_t) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_size(arg, scheme_name, idx);
+	}
+	int scm_to(SCM args, size_t idx, int) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_int(arg, scheme_name, idx);
+	}
+	double scm_to(SCM args, size_t idx, double) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_real(arg, scheme_name, idx);
+	}
+	std::string scm_to(SCM args, size_t idx, const std::string&) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_string(arg, scheme_name, idx);
+	}
+	Type scm_to(SCM args, size_t idx, Type) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_atom_type(arg, scheme_name, idx);
+	}
+	Handle scm_to(SCM args, size_t idx, const Handle&) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_handle(arg, scheme_name, idx);
+	}
+	AtomSpace* scm_to(SCM args, size_t idx, const AtomSpace*) const
+	{
+		SCM arg = scm_list_ref(args, scm_from_size_t(idx));
+		return SchemeSmob::verify_atomspace(arg, scheme_name, idx);
```

Comment:
this should probably be removed, no one should be passing atomspaces.  Again, another abuse of the API.